pool/heap.go

```diff
@@ -0,0 +1,218 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package pool
+
+import (
+	"reflect"
+	"sort"
+	"strconv"
+	"sync"
+	"time"
+	"unsafe"
+
+	"github.com/uber-go/tally"
+)
+
+// NewNativeHeap constructs a new BytesPool based on NativePool.
+func NewNativeHeap(b []Bucket, po ObjectPoolOptions) BytesPool {
+	if po == nil {
+		po = NewObjectPoolOptions()
+	}
+
+	var (
+		m        = po.MetricsScope()
+		ByteType = reflect.TypeOf((byte)(0))
+	)
+
+	sort.Sort(BucketByCapacity(b))
+
+	r := heap{m: heapMetrics{
+		overflows: m.Counter("overflows"),
+		misplaces: m.Counter("misplaces"),
+	}}
+
+	for _, cfg := range b {
+		m := m.Tagged(map[string]string{
+			"class": strconv.Itoa(cfg.Capacity),
+		})
+
+		v := &slot{class: cfg.Capacity, opts: NativePoolOptions{
+			Size: uint(cfg.Count),
+			Type: reflect.ArrayOf(cfg.Capacity, ByteType),
+		}, m: slotMetrics{
+			used: m.Gauge("used"),
+			size: m.Gauge("size"),
+		}}
+
+		r.slots = append(r.slots, v)
+	}
+
+	return r
+}
+
+type heap struct {
+	slots []*slot
+
+	m heapMetrics
+}
+
+type heapMetrics struct {
+	overflows tally.Counter
+	misplaces tally.Counter
+}
+
+type slot struct {
+	sync.RWMutex
+
+	class int
+	opts  NativePoolOptions
+	pools []NativePool
+
+	m slotMetrics
+}
+
+type slotMetrics struct {
+	used tally.Gauge
+	size tally.Gauge
+}
+
+func (s *slot) get() interface{} {
+	if segment := s.getOr(s.RLocker(), func() interface{} {
+		return nil
+	}); segment != nil {
+		return segment
+	}
+
+	// Slow path - double-check that there are no segments left,
+	// then grow while holding an exclusive lock.
+	return s.getOr(s, func() interface{} {
+		s.pools = append(
+			[]NativePool{NewNativePool(s.opts)}, s.pools...)
+		return s.pools[0].Get()
+	})
+}
+
+func (s *slot) getOr(l sync.Locker, fn OverflowFn) interface{} {
+	var segment interface{}
+
+	l.Lock()
+
+	for _, pool := range s.pools {
+		if segment = pool.GetOr(func() interface{} {
+			return nil
+		}); segment != nil {
+			l.Unlock()
+			s.updateMetrics()
+			return segment
+		}
+	}
+
+	segment = fn()
+
+	l.Unlock()
+
+	return segment
+}
+
+func (s *slot) put(segment interface{}) {
+	s.RLock()
+
+	for i := range s.pools {
+		if s.pools[i].Owns(segment) {
+			s.pools[i].Put(segment)
+			break
+		}
+	}
+
+	s.RUnlock()
+}
+
+func (s *slot) updateMetrics() {
+	// TODO(@kobolog): Use Dice.
+	if time.Now().UnixNano()%sampleObjectPoolLengthEvery != 0 {
+		return
+	}
+
+	var used, size int64
+
+	s.RLock()
+
+	for i := range s.pools {
+		a, b := s.pools[i].Size()
+
+		used += int64(a)
+		size += int64(b)
+	}
+
+	s.RUnlock()
+
+	s.m.used.Update(used)
+	s.m.size.Update(size)
+}
+
+func (p heap) Init() {
+	for _, slot := range p.slots {
+		slot.pools = append(slot.pools, NewNativePool(slot.opts))
```

Comment:
I think it would be cleaner if `slot` had it's own `init` method that sets the `pools` property.  That way any other initialization that may need to be added in the future can be encapsulated well and the responsibility to initialize fields is delegated to the `slot` itself.